### ISSUE TITLE
docs: Document that Angular relies on utf8 files

### ISSUE THREAD

**Issue by SnareChops** [angular/angular#15638]
**I'm submitting a ...**
```
[x] bug report
[ ] feature request
[ ] support request
```
If I believed this to be the cause of user error then I would post on StackOverflow. But with how basic this reproduction is and with previous bugs reported with the same error message and went unresolved due to no reproductions available, it seemed best to submit this here with a full reproduction example.

**Current behavior**
Application errors during bootstrap as shown below with a barebones angular application

**Expected behavior**
Application should bootstrap and display

**Minimal reproduction of the problem with instructions**
https://github.com/SnareChops/OverwatchAlgebra
```
npm install
npm run build:site
serve / open index.html in browser of your choosing using server of your choosing
```

**Please tell us about your environment**
Mac OSX (10.12.3)
Rollup IIFE
Node 7.5.0

Angular version: 4.0.1

Browser: chrome (haven't tested others)

Language: Typescript 2.2.1

## Error received
<img width="1205" alt="screen shot 2017-03-30 at 4 04 20 pm" src="https://cloud.githubusercontent.com/assets/3210560/24523871/92c9a25a-1562-11e7-8090-48b57cec76c8.png">

Application is barebones. Entire application code is:

```
@Component({
	selector: 'owa',
	template: `
		<p>Overwatch Algebra</p>
	`
})
export class MainComponent{}

@NgModule({
	declarations: [
		MainComponent
	],
	imports: [
		CommonModule,
		BrowserModule
	],
	schemas: [
		CUSTOM_ELEMENTS_SCHEMA
	],
	bootstrap: [
		MainComponent
	]
})
export class MainModule{}

platformBrowser().bootstrapModuleFactory(MainModuleNgFactory);
```
(of course in separate files and with relevant imports)

**Comment by jponeill** [comment]
Not sure it's related, but for me this was a character set issue and switching from ISO-8859-1 to UTF-8 fixed the problem. Lines like the following in the "main" bundle were causing problems (e.g, "_\u0275...")

Object.defineProperty(AppModuleInjector.prototype, "_\u0275AnimationEngine_93", {
    get: function () {
        if ((this.__ɵAnimationEngine_93 == null)) {
            (this.__ɵAnimationEngine_93 = new __WEBPACK_IMPORTED_MODULE_34__angular_platform_browser_animations__["c" /* ɵa */](this._AnimationDriver_91, this._ɵAnimationStyleNormalizer_92));

**Comment by tbosch** [comment]
Yes, angular requires utf8, sorry.

**Comment by Quantas** [comment]
@tbosch Thanks for the update, out of curiosity, is this documented anywhere? I ask because apps worked before 4.0.0-beta.7 and then suddenly stopped working.

**Comment by tbosch** [comment]
Mmh, good point. Reopening so we update the docs...

**Comment by SnareChops** [comment]
I ran the following command on my project folder to find all non UTF-8 files
```
find . -type f -exec file --mime {} \; | grep -v '.git' | grep -v "charset=utf-8"
```

And I found that all of angular's files are ASCII and mine as well. Is this the issue? If so why would angular publish the files in ASCII if UTF-8 is required? Should this really matter since UTF-8 extends ASCII?

```
./node_modules/@angular/common/@angular/common/testing.es5.js: text/plain; charset=us-ascii
./node_modules/@angular/common/@angular/common/testing.es5.js.map: text/plain; charset=us-ascii
./node_modules/@angular/common/@angular/common/testing.js: text/x-c++; charset=us-ascii
./node_modules/@angular/common/@angular/common/testing.js.map: text/plain; charset=us-ascii
./node_modules/@angular/common/bundles/common-testing.umd.js: text/plain; charset=us-ascii
./node_modules/@angular/common/bundles/common-testing.umd.js.map: text/plain; charset=us-ascii
./node_modules/@angular/common/bundles/common-testing.umd.min.js: text/plain; charset=us-ascii
./node_modules/@angular/common/bundles/common-testing.umd.min.js.map: text/plain; charset=us-ascii
./node_modules/@angular/common/package.json: text/plain; charset=us-ascii
./node_modules/@angular/common/public_api.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/README.md: text/plain; charset=us-ascii
./node_modules/@angular/common/src/common_module.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/directives/index.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/directives/ng_class.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/directives/ng_component_outlet.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/directives/ng_for_of.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/directives/ng_if.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/directives/ng_plural.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/directives/ng_style.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/directives/ng_switch.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/directives/ng_template_outlet.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/localization.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/location/hash_location_strategy.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/location/index.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/location/location.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/location/location_strategy.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/location/path_location_strategy.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/location/platform_location.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/pipes/async_pipe.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/pipes/case_conversion_pipes.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/pipes/date_pipe.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/pipes/i18n_plural_pipe.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/pipes/i18n_select_pipe.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/pipes/index.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/pipes/intl.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/pipes/invalid_pipe_argument_error.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/pipes/json_pipe.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/pipes/number_pipe.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/pipes/slice_pipe.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/platform_id.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/src/version.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/testing/index.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/testing/index.metadata.json: text/plain; charset=us-ascii
./node_modules/@angular/common/testing/package.json: text/plain; charset=us-ascii
./node_modules/@angular/common/testing/src/location_mock.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/testing/src/location_mock.metadata.json: text/plain; charset=us-ascii
./node_modules/@angular/common/testing/src/mock_location_strategy.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/testing/src/mock_location_strategy.metadata.json: text/plain; charset=us-ascii
./node_modules/@angular/common/testing/src/testing.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/testing/src/testing.metadata.json: text/plain; charset=us-ascii
./node_modules/@angular/common/testing.d.ts: text/plain; charset=us-ascii
./node_modules/@angular/common/testing.metadata.json: text/plain; charset=us-ascii
... (cut list for brevity, but ALL @angular/* files are charset=us-ascii

./package.json: text/plain; charset=us-ascii
./README.md: text/plain; charset=us-ascii
./rollup-config.js: text/plain; charset=us-ascii
./src/components/main.component.ts: text/plain; charset=us-ascii
./src/index-jit.ts: text/plain; charset=us-ascii
./src/index.ts: text/plain; charset=us-ascii
./tsconfig.json: text/plain; charset=us-ascii
```

**Comment by SnareChops** [comment]
@tbosch 

I've done further research into this and have found the untranslated unicode characters that @jponeill posted. For example my bundle contained

```
Object.defineProperty(MainModuleInjector.prototype, "_\u0275DomRendererFactory2_21", {
        ...
    });
```

which is the reason for the error in the DI system. `injector.get(RendererFactory2)` was returning undefined. Therefore `this.delegate` in the `DebugRendererFactory2.createRenderer()` method was failing. So now we've identified the cause of the error, however this is not the fault of non UTF-8 files, or the bundler. All files are UTF8 all the way through the process as I believe I have proven above, and the DomRendererFactory2 is a private export using your `ɵDomRendererFactory2` convention.

In the `index.ngfactory.js` file that is generated **from the compiler-cli** the same translation error occurs

```
Object.defineProperty(MainModuleInjector.prototype, "_\u0275DomRendererFactory2_21", {
        ...
    })
```

This is **before** any bundling or any of my scripts touch any files. So this is coming directly from an angular lib or process.

This appears to be an issue with the AOT compiler-cli incorrectly `.toString()` on the `name` of the class. To try and resolve this issue I created a node script that I run on the bundle **after** the entire build process.

```
#!/usr/bin/node
const fs = require('fs');

let data = fs.readFileSync('bundle.js');
data = data.toString().replace(/\\u0275/g, 'ɵ');
fs.writeFileSync('bundle.js', data);
```

This corrects the erroneous output from the AOT compiler and fixes the issue. The code then runs exactly as expected without any problems.

**Comment by DanielSchaffer** [comment]
My app is still failing with the same error as above despite running the replace fix above. Hashing the "data" var before and after the fix produces different hashes, so it's doing _something_, it's just not fixing it.

Also of note, I only have this problem when I import the `ngx-color-picker` module (from https://github.com/zefoy/ngx-color-picker). Removing that dependency allows the app to run normally.

**Comment by DanielSchaffer** [comment]
Turns out, I was just missing a `<meta charset="utf-8">` tag. I don't even need the replace fix with the tag in place.

**Comment by wardbell** [comment]
I wonder where we should put this info so people can find it. "Setup"? We don't talk about `index.html`.

**Comment by wardbell** [comment]
After discussion and reflection, I decided that this is a good candidate for Stack Overflow but we shouldn't document this in Angular docs for several reasons:

1) This is not really about Angular _per se_. It's a general web issue
2) utf8 is standard across the web, not just in Angular apps
3) We shouldn't have to tell people to add `<meta charaset+"utf-8">` to `index.html` (it's in all of our samples and the CLI seed)
4) We shouldn't have to tell people to configure their web server for utf8
5) Even if we did add this to the setup page, almost no one would find it. The error message doesn't draw your attention to it.
6) This kind of advice adds unfairly to the perceived complexity of Angular because it makes the instructions and warnings bigger. There just isn't enough value in this particular piece of advice to compensate.
